### PR TITLE
Refactor(ES6): DelegatedModule.js

### PR DESCRIPTION
**Pull request by sergiorojasa**
**What kind of change does this PR introduce?**
Refactor, ES5 => ES6

**Did you add tests for your changes?**
All existing tests pass

**If relevant, link to documentation update:**
N/A

**Summary**
Refactor DelegatedModule.js

**Does this PR introduce a breaking change?**
No

### CODE REVIEW

**Comment by sokra** [comment]
`Uncaught TypeError: this.disconnect is not a function`

File: lib/DelegatedModule.js [hunk 1]
```diff
@@ -2,79 +2,80 @@
...
+const OriginalSource = require("webpack-sources").OriginalSource;
+const RawSource = require("webpack-sources").RawSource;
+const WebpackMissingModule = require("./dependencies/WebpackMissingModule");
+const DelegatedSourceDependency = require("./dependencies/DelegatedSourceDependency");
 
-DelegatedModule.prototype = Object.create(Module.prototype);
-DelegatedModule.prototype.constructor = DelegatedModule;
+class DelegatedModule {
+	constructor(sourceRequest, data, type, userRequest) {
+		Module.call(this);
```
Comment (Kovensky) [inline lib/DelegatedModule.js]: While `Module` has not yet been converted to `class` syntax, it is a class. _Never_ call a class as a function, they _will_ throw even if it's an ES5 construction pattern.

The DelegatedModule's prototype's prototype is also the `Module` prototype so this should be:

```js
class DelegatedModule extends Module {
  constructor(sourceRequest, data, type, userRequest) {
    super()
    /* ... */
  }
  /* ... */
}
```

File: lib/DelegatedModule.js [hunk 2]
```diff
@@ -2,79 +2,80 @@
...
-};
 
-DelegatedModule.prototype.size = function() {
-	return 42;
-};
+	size() {
+		return 42;
+	}
+}
+DelegatedModule.delegated = true;
```
Comment (Kovensky) [inline lib/DelegatedModule.js]: This was a `prototype` property, not a constructor (`static`) property. It either should still be `DelegatedModule.prototype.delegated = true`, or should be `this.delegated = true` in the constructor to make it own.

File: lib/DelegatedModule.js [hunk 3]
```diff
@@ -2,79 +2,80 @@
...
+		this.meta = data.meta;
+		this.type = type;
+		this.userRequest = userRequest;
+		this.built = false;
+		this.delegateData = data;
+	}
 
-DelegatedModule.prototype.delegated = true;
+	identifier() {
+		return "delegated " + JSON.stringify(this.request) + " from " + this.sourceRequest;
```
Comment (Kovensky) [inline lib/DelegatedModule.js]: This should be a template string

File: lib/DelegatedModule.js [hunk 4]
```diff
@@ -2,79 +2,80 @@
...
+	}
 
-DelegatedModule.prototype.identifier = function() {
-	return "delegated " + JSON.stringify(this.request) + " from " + this.sourceRequest;
-};
+	readableIdentifier() {
+		return "delegated " + this.userRequest + " from " + this.sourceRequest;
```
Comment (Kovensky) [inline lib/DelegatedModule.js]: same

File: lib/DelegatedModule.js [hunk 5]
```diff
@@ -2,79 +2,80 @@
...
+		this.addDependency(new DelegatedSourceDependency(this.sourceRequest));
+		callback();
+	}
 
-DelegatedModule.prototype.needRebuild = function() {
-	return false;
-};
+	unbuild() {
+		this.built = false;
+		Module.prototype.unbuild.call(this);
```
Comment (Kovensky) [inline lib/DelegatedModule.js]: `super.unbuild()` (after fixing the class declaration)

File: lib/DelegatedModule.js [hunk 6]
```diff
@@ -2,79 +2,80 @@
...
+			}
+		}
 
-DelegatedModule.prototype.source = function() {
-	var sourceModule = this.dependencies[0].module;
-	var str;
-	if(!sourceModule) {
-		str = WebpackMissingModule.moduleCode(this.sourceRequest);
-	} else {
-		str = "module.exports = (__webpack_require__(" + sourceModule.id + "))";
```
Comment (Kovensky) [inline lib/DelegatedModule.js]: This line is missing in the rewrite; the `str +=` lines will append to `undefined`.

File: lib/DelegatedModule.js [hunk 7]
```diff
@@ -2,79 +2,80 @@
...
-};
+		if(!sourceModule) {
+			str = WebpackMissingModule.moduleCode(this.sourceRequest);
+		} else {
+			switch(this.type) {
+				case "require":
+					str += "(" + JSON.stringify(this.request) + ");";
+					break;
+				case "object":
+					str += "[" + JSON.stringify(this.request) + "];";
```
Comment (Kovensky) [inline lib/DelegatedModule.js]: These also should be template strings

**Comment by sergiorojasa** [comment]
@Kovensky all pass but 1

File: lib/DelegatedModule.js [hunk 8]
```diff
@@ -2,79 +2,81 @@
...
+const RawSource = require("webpack-sources").RawSource;
+const WebpackMissingModule = require("./dependencies/WebpackMissingModule");
+const DelegatedSourceDependency = require("./dependencies/DelegatedSourceDependency");
 
-DelegatedModule.prototype = Object.create(Module.prototype);
-DelegatedModule.prototype.constructor = DelegatedModule;
+class DelegatedModule extends Module {
+	constructor(sourceRequest, data, type, userRequest) {
+		super();
+		Module.call(this);
```
Comment (sokra) [inline lib/DelegatedModule.js]: remove this line. `Module.call(this);` is `super();`

File: lib/DelegatedModule.js [hunk 9]
```diff
@@ -2,79 +2,81 @@
...
-	this.builtTime = new Date().getTime();
-	this.usedExports = true;
-	this.providedExports = this.delegateData.exports || true;
-	this.dependencies.length = 0;
-	this.addDependency(new DelegatedSourceDependency(this.sourceRequest));
-	callback();
-};
+	unbuild() {
+		this.built = false;
+		Module.prototype.unbuild.call(this);
```
Comment (sokra) [inline lib/DelegatedModule.js]: Use the `super` syntax.

**Comment by sokra** [comment]
already merged

**Pull request closed without merging**